### PR TITLE
Adding A new Component

### PR DESCRIPTION
**Pull request by Shivamraghu1208**

### CODE REVIEW

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import java.util.ArrayList;
+
+@Model(adaptables = Resource.class, defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL)
+public class ProductDetailModel {
+
+    @OSGiService
+    MoviesService moviesService;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private.

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    @OSGiService
+    ProductDetailService productDetailService;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @ValueMapValue
+    String heading;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    private Logger logger = LoggerFactory.getLogger(ProductDetailModel.class);
+
+    public String[] getMovies() {
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: move this method after below the variables declaration.

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        return movies;
+    }
+
+    private String[] movies;
+
+    String response;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    public ArrayList<Product> getList() {
+        return list;
+    }
+
+    ArrayList<Product> list = new ArrayList<>();
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    public String getResponse() {
+        return response;
+    }
+
+
+    @ValueMapValue
+    String select;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+    @ValueMapValue
+    long NumOfProducts;
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: pls mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+
+    @PostConstruct
+    protected void init() {
+
+        response = productDetailService.getResponse();
+        JsonObject jsonObject = JsonParser.parseString(response).getAsJsonObject();
+        if (!jsonObject.isJsonNull()) {
+            if (!jsonObject.get("products").isJsonNull()) {
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: merge if statements on line number 72 & 73

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java (hunk 1 above)
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: No use of this moviesService variable, please remove this.

File: core/src/main/java/com/adobe/aem/sample/site/core/services/impl/ProductDetailServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public String getResponse() {
+        return response;
+    }
+
+    @Activate
+    @Modified
+    protected void activate(ProductDetailServiceConfiguration configuration){
+        String s = configuration.apiUrl();
```
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/services/impl/ProductDetailServiceImpl.java]: Variable name should be meaning full.

File: core/src/main/java/com/adobe/aem/sample/site/core/services/impl/MoviesServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.metatype.annotations.Designate;
+
+@Component(service = MoviesService.class, immediate = false)
+@Designate(ocd= MoviesServiceConfiguration.class)
+public class MoviesServiceImpl implements MoviesService {
+
+    String[] str;
```
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/services/impl/MoviesServiceImpl.java]: Variable name should be meaningfull

File: ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                            emptyText="Enter Product Heading"
+                            fieldLabel=" Product Heading"
+                            name="./heading"
+                            required="{Boolean}true"/>
+
+                     <select
+                            jcr:primaryType="nt:unstructured"
+                            sling:resourceType="granite/ui/components/coral/foundation/form/select"
+                            fieldLabel="Choose Size"
+                            name="./select">
```
Comment (akashsen-aurx) [inline ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml]: Give more meaning full name

File: ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                                    jcr:primaryType="nt:unstructured"
+                                    text="h4 "
+                                    value="h4"/>
+                            <h5
+                                    jcr:primaryType="nt:unstructured"
+                                    text="h5 "
+                                    value="h5"/>
+                            <h6
+                                    jcr:primaryType="nt:unstructured"
+                                    text="h6 tag"
```
Comment (akashsen-aurx) [inline ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml]: update this property

File: ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+                                    value="h6"/>
+                        </items>
+                    </select>
+                    <Num
+                            jcr:primaryType="nt:unstructured"
+                            sling:resourceType="granite/ui/components/coral/foundation/form/numberfield"
+                            emptyText="Enter No of Products"
+                            fieldLabel="Number Of Products"
+                            name="./NumOfProducts"
```
Comment (akashsen-aurx) [inline ui.apps/src/main/content/jcr_root/apps/aemtraining/components/productdetail/_cq_dialog/.content.xml]: Provide meaning full Name

File: ui.apps/src/main/content/jcr_root/apps/aemtraining/components/structure/StudentPage/customfooterlibs.html [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+
+<!--/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Create overlay of this file to include custom client libs in the page footer
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+
+<sly data-sly-use.clientlib="core/wcm/components/commons/v1/templates/clientlib.html">
+    <sly data-sly-call="${clientlib.js @ categories='employee-clientlib'}"/>
```
Comment (akashsen-aurx) [inline ui.apps/src/main/content/jcr_root/apps/aemtraining/components/structure/StudentPage/customfooterlibs.html:22]: What is the need of this clientlibs in page

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    MoviesService moviesService;
+    @OSGiService
+    ProductDetailService productDetailService;
+    @ValueMapValue
+    String heading;
+
+    private Logger logger = LoggerFactory.getLogger(ProductDetailModel.class);
+
+    public String[] getMovies() {
+        return movies;
```
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: Are you using this getter anywhere, if not then remove

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java (hunk 6 above)
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: Update variable name

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+                        JsonElement jsonElement = jsonArray.get(i);
+
+                        JsonObject asJsonObject = jsonElement.getAsJsonObject();
+                        int id = asJsonObject.get("id").getAsInt();
+                        String title = asJsonObject.get("title").getAsString();
+                        String brand = asJsonObject.get("brand").getAsString();
+                        long price = asJsonObject.get("price").getAsLong();
+                        JsonArray images = asJsonObject.get("images").getAsJsonArray();
+                        String image = images.get(0).getAsString();
+                        Product p = new Product();
```
Comment (akashsen-aurx) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: Where is the Product class in project

**Review by akashsen-aurx — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/adobe/aem/sample/site/core/services/impl/MoviesServiceImpl.java (hunk 11 above)
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/services/impl/MoviesServiceImpl.java]: mark this as private

File: core/src/main/java/com/adobe/aem/sample/site/core/services/impl/ProductDetailServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        try {
+          URL url = new URL(apiUrl);
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setRequestProperty("accept", "application/json");
+            InputStream inputStream = connection.getInputStream();
+            response = IOUtils.toString(inputStream, StandardCharsets.UTF_8);
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/services/impl/ProductDetailServiceImpl.java]: why printstacktrace, add a logger

**Review by abhinav-agarwal — CHANGES_REQUESTED** [review]

**Comment by Shivamraghu1208** [comment]
Hello sir,

Requested changes updated.

________________________________

File: core/src/main/java/com/adobe/aem/sample/site/core/services/config/ProductDetailServiceConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package com.adobe.aem.sample.site.core.services.config;
+
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+
+@ObjectClassDefinition(name="Product detail configuration")
+public @interface ProductDetailServiceConfiguration {
+    @AttributeDefinition(name="ApiUrl")
+    String apiUrl();
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/services/config/ProductDetailServiceConfiguration.java]: this should be api_url

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java (hunk 6 above)
Comment (abhinav-agarwal) [reply core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: still not private

File: core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
```
Comment (abhinav-agarwal) [inline core/src/main/java/com/adobe/aem/sample/site/core/models/ProductDetailModel.java]: please use a meaningful variable name

**Review by abhinav-agarwal — CHANGES_REQUESTED** [review]

**Review by abhinav-agarwal — APPROVED** [review]

**Pull request merged**